library/core/src/mem/mod.rs

```diff
@@ -642,24 +652,19 @@ pub unsafe fn zeroed<T>() -> T {
 /// is special in that the compiler knows that it does not have a fixed value.
 /// This makes it undefined behavior to have uninitialized data in a variable even
 /// if that variable has an integer type.
-/// (Notice that the rules around uninitialized integers are not finalized yet, but
-/// until they are, it is advisable to avoid them.)
 ///
 /// [uninit]: MaybeUninit::uninit
 /// [assume_init]: MaybeUninit::assume_init
 /// [inv]: MaybeUninit#initialization-invariant
 #[inline(always)]
 #[rustc_deprecated(since = "1.39.0", reason = "use `mem::MaybeUninit` instead")]
 #[stable(feature = "rust1", since = "1.0.0")]
-#[allow(deprecated_in_future)]
-#[allow(deprecated)]
 #[rustc_diagnostic_item = "mem_uninitialized"]
 pub unsafe fn uninitialized<T>() -> T {
     // SAFETY: the caller must guarantee that an unitialized value is valid for `T`.
-    unsafe {
-        intrinsics::assert_uninit_valid::<T>();
```

Comment:
Can you leave this call in place? That will allow us to keep the "attempted to leave type `T` uninitialized, which is invalid" panics.

Reply:
I was under the impression from your previous comments that you were content with only panicking on can't-be-zero types. If your concern is just the diagnostic regression, I would much rather solve this some other way than reintroducing this check, since one of the points of this PR is to deliberately make it possible to use this function with types like `bool`, so as to "un-break" code that was broken when these checks were added in the first place.

Reply:
I think I may have misunderstood what was happening.


These panics were intentionally added in https://github.com/rust-lang/rust/pull/66059, as a way of catching misuses of uninitialized memory. Even though `std::mem::uninitialized` now 'happens' to zero out these types, users **should not** be relying on this behavior (as your doc comment states). That is, writing `std::mem::uninitialized::<bool>()` is still a bug - any code doing so is broken, and should not be 'un-broken' by removing the legitimate panic.

Reply:
@Aaron1011 I thought the whole point was that `mem::uninitialized` is broken for *any* type. Why are types with niches being singled out?

Reply:
See also @thomcc's comments about how the panics have broken several of his old crates; I would *love* for this to fix them. https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/disarming.20mem.3A.3Auninitialized/near/244037069

Reply:
Note that this is the only user of that function, so if we end up removing it we might also want to remove its support infrastructure.

> I thought the whole point was that mem::uninitialized is broken for any type. Why are types with niches being singled out?

It is "slightly less broken" for types like `i32` in the sense that, as far as we know, it will not actually cause miscompilations when using with such a type.

It is in fact not at all broken for types that are allowed to be uninitialized, such as `MaybeUninit<T>`.

> That is, writing std::mem::uninitialized::<bool>() is still a bug - any code doing so is broken, and should not be 'un-broken' by removing the legitimate panic.

I guess the point is that yes, it is a bug, but if we can guarantee that nothing will actually go horribly wrong then maybe we should also not have the panic -- therefore we increase "de-facto" but not "de-jure" compatibility with old broken code.

This brings me to an idea... can you make this code still return proper uninit memory under `cfg(miri)`? That way Miri can still detect incorrect uses of this function.

Reply:
That said, this check (`assert_uninit_valid`) is currently still fairly conservative, see https://github.com/rust-lang/rust/issues/66151. So I wonder if it would not be better to crank up that check before removing it, so that we get more people to actually fix their code.

This was started in https://github.com/rust-lang/rust/pull/79296 but then stalled when the rather large crater results had to be analyzed.

Reply:
> It is in fact not at all broken for types that are allowed to be uninitialized, such as MaybeUninit<T>.

Are there any types that are allowed to be uninitialized other than MaybeUninit?

Reply:
> This was started in #79296 but then stalled when the rather large crater results had to be analyzed.

I queued a newer Crater run in https://github.com/rust-lang/rust/pull/87041, but it will be a while before it gets to the head of the queue and finished.

Reply:
> Are there any types that are allowed to be uninitialized other than MaybeUninit (and arrays and tuples of MaybeUninit)?

Yes:
- The type `()`
- Any user-defined `union` type... well to be fair this is not ratified yet, that's just what I would propose. But I think there is consensus that any `union` with a field of size zero (essentially, a user defining their own `MaybeUninit`) is okay to be left uninitialized.

Now, I am not saying that this is all that useful, I just want to get our facts straight so we have a solid basis to make decisions on. :) . The only legitimate reason I can think of to call `mem::uninitialized` is to create an array of `MaybeUninit`, and we should probably just offer a better API for that (possibly using const generics, like other stdlib APIs are already doing AFAIK).

Reply:
That lint however has a lot of false negatives. For example, it will not fire when the code that calls `mem::uninitialized` is generic.